### PR TITLE
reimplement `conditional_seal` in terms of `seal`

### PR DESCRIPTION
**Pull request by ruchirK**
### Motivation

See individual commits for details.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by ruchirK** [comment]
RFAL! @aljoscha I'm really curious what you think here -- in particular I'm not thrilled by the second commit and really tried to make `seal` pass the "advances only only success" test with the `build_async` version, but just couldn't get it to work.

**Comment by aljoscha** [comment]
Ha! I recently thought about dropping `seal()` because it's a) not correct, and b) we only use `conditional_seal()` but this way round also seems interesting!

We probably leave some performance on the table by relying on the timely progress tracking machinery instead of doing it the "optimal way" ourselves, in just one operator. But that might be negligible.

On implementing using `build_async`: I have a feeling that might be hard because a) our `PFutures` are different in that they don't require polling to finish (because they run on the persist thread pool). And we want to potentially have multiple futures in flight, which seems harder to express in an async version, where you would then have to use combinators to combine all the futures and poll them together.

**Comment by ruchirK** [comment]
Yeah I agree that this approach probably leaves a bit of control plane performance on the table -- but i think at this stage its absolutely a worthwhile tradeoff to make, in return for a much simpler primitive (`seal` vs `conditional_seal`), and fewer custom operators. I'll benchmark this more today to see what the perf hit is like / if its even measurable

**Comment by ruchirK** [comment]
ok this is all ready to go!

**Comment by ruchirK** [comment]
a very very gentle ping @aljoscha

**Comment by aljoscha** [comment]
Oh boy, sorry I dropped the ball on this one. Reviewing now.

File: src/persist/src/operators/stream.rs [hunk 1]
```diff
@@ -265,58 +264,106 @@ where
...
+                            log::trace!(
+                                "In {}, finished sealing collection up to {}",
+                                &operator_name,
+                                pending_future.time,
+                            );
+                            // Explicitly downgrade the capability to the new time.
+                            cap_set.downgrade(Some(pending_future.time));
+                        }
+                        std::task::Poll::Ready(Err(e)) => {
+                            let error = format!(
```
Comment (aljoscha) [inline src/persist/src/operators/stream.rs:314]: nit: This is a remnant from my code, when it was still both emitting the error and logging it. Now you could just put this directly in the `log::trace!()` call.

**Review by aljoscha — APPROVED** [review]
It's very nice how this consolidates the code! 🙌

**Comment by ruchirK** [comment]
No worries and thanks for the review! Merging on green :D

**Pull request merged**